src/Expander/Monad.hs

```diff
@@ -267,9 +268,10 @@ data ExpanderState = ExpanderState
   , _expanderTasks :: [(TaskID, ExpanderLocal, ExpanderTask)]
   , _expanderOriginLocations :: !(Map.Map SplitCorePtr SrcLoc)
   , _expanderCompletedCore :: !(Map.Map SplitCorePtr (CoreF TypePatternPtr PatternPtr SplitCorePtr))
-  , _expanderCompletedPatterns :: !(Map.Map PatternPtr ConstructorPattern)
+  , _expanderCompletedPatterns :: !(Map.Map PatternPtr (ConstructorPatternF PatternPtr))
   , _expanderCompletedTypePatterns :: !(Map.Map TypePatternPtr TypePattern)
-  , _expanderPatternBinders :: !(Map.Map (Either PatternPtr TypePatternPtr) [(Scope, Ident, Var, SchemePtr)])
+  , _expanderPatternBinders :: !(Map.Map PatternPtr (Either [PatternPtr] (Scope, Ident, Var, SchemePtr)))
+  , _expanderTypePatternBinders :: !(Map.Map TypePatternPtr [(Scope, Ident, Var, SchemePtr)])
```

Comment:
So when pattern-matching on types, we must still do it a single layer at a time, presumably because of the interaction with the getting-stuck mechanism? If so, it seems like we will still end up writing a macro desugaring deep type-pattern matching into shallow type-pattern matching. But since that macro is manipulating syntax, not types, I guess we don't have the same boot-strapping problem?

Reply:
Matching on types is still a monadic action due to the capturing of the continuation.

If we make the structure of the captured continuations more complicated, then we can support deep matching on types directly, and I think this would probably be a nice improvement. But it's not in this PR.